[BasicBlockSections] Introduce the basic block sections profile version 1.

rlavaee opened the pull request:
This patch introduces a new version for the basic block sections profile as was requested in D158442, while keeping backward compatibility for the old version.

The new encoding is as follows:
```
m <module_name>
f <function_name_1> <function_name_2>...
c <bb_id_1> <bb_id_2> <bb_id_3>
c <bb_id_4> <bb_id_5>
...
```
Module name specifier (starting with 'm') is optional and allows distinguishing profiles for internal-linkage functions with the same name. If not specified, profile will be applied to any function with the same name.
Function name specifier (starting with 'f') can specify multiple function name aliases.
Finally, basic block clusters are specified by 'c' and specify the cluster of basic blocks, and the internal order in which they must be placed in the same section.

shenhanc78 reviewed (commented):
`f <function name 1> <function name 2>`
So func name 1 and func name 2 are aliaes, right?

HazardyKnusperkeks commented:
There went something wrong, hm?

rlavaee committed:
[BasicBlockSections] Introduce the basic block sections profile version 1. #65506

This patch introduces a new version for the basic block sections profile as was requested in D158442, while keeping backward compatibility for the old version.

The new encoding is as follows:

m <module_name>
f <function_name_1> <function_name_2>...
c <bb_id_1> <bb_id_2> <bb_id_3>
c <bb_id_4> <bb_id_5>
...
Module name specifier (starting with 'm') is optional and allows distinguishing profiles for internal-linkage functions with the same name. If not specified, profile will be applied to any function with the same name.
Function name specifier (starting with 'f') can specify multiple function name aliases.
Finally, basic block clusters are specified by 'c' and specify the cluster of basic blocks, and the internal order in which they must be placed in the same section.

asl reviewed (dismissed):
Consider rebasing your commit before submission.

rlavaee committed:
Add some comments.

rlavaee commented:
> `f <function name 1> <function name 2>` So func name 1 and func name 2 are aliaes, right?

Correct. Fixed the function comment.

rlavaee commented:
ping. @tmsri

rlavaee commented:
Another friendly ping for review.

tmsri commented on llvm/lib/CodeGen/BasicBlockSectionsProfileReader.cpp:
```diff
@@ -46,34 +47,121 @@ BasicBlockSectionsProfileReader::getBBClusterInfoForFunction(
+// aliases. Basic block clusters are specified by 'c' and specify the cluster of
+// basic blocks, and the internal order in which they must be placed in the same
+// section.
+Error BasicBlockSectionsProfileReader::ReadV1Profile() {
```
We have to put an expiry date on "ReadV0Profile".  create_llvm_prof has to be updated to start producing the new format and we should obsolete and delete the old function soon-ish.

rlavaee committed:
Add comment for deprecation.

rlavaee committed:
clang-format.
